### DISCUSSION TITLE
How to generate thumbnails for uploaded video?

### DISCUSSION THREAD

**Discussion by piszczu4** [vidstack/player#1017] [Q&A] [answered]
Hi let's say that I uploaded video and I have it's URL. How can I easily generate thumbnails file for this video so that I can pass its URL to the player instance? Similarly for Poster?

**Comment by mihar-22** [comment, marked as answer]
The easy way is to use an existing service to host/stream your videos and they will provide simple APIs to do so such as [Mux](https://www.mux.com/) or [Cloudflare Stream](https://developers.cloudflare.com/stream/). These services are just my favourites but there's tons more options out there you can find.

The "harder" way depending on how comfortable you are with CLI and [FFMPEG](https://ffmpeg.org/) is to generate them yourself. You can use ChatGPT to help you build the commands but it usually takes some tweaking to get it right. Once you have the thumbnails sprite, you'll need to build the VTT file which will also require some back and forth with ChatGPT or searching the web.

Hope that helps! Not ideal but I hope we can make this all easier in the future through Vidstack.

<img width="799" alt="image" src="https://github.com/vidstack/player/assets/14304599/a597261a-09f3-44f2-a948-e659621da547">

**Reply by piszczu4** [reply]
Hi, thank you, very helpful. Currently I'm using Digital Ocean Spaces to store my videos but the drawback is that I can't actually let user to choose the quality. If I'd like to use MUX, do I have to move all my videos there? I choose DO Spaces over for example vimeo to be not limited by the number of videos I could upload. So in the end I want to use vidstack player and store the videos in such a way that I can easily generate thumbnails and let user to pick the quality. What setup would you suggest me?

**Reply by mihar-22** [reply]
Both Mux and Cloudflare Stream are super affordable now. Encoding on Mux is free, so you'll simply pay for really cheap storage and streaming (for up to 1080p). Both of those services will give you a HLS playlist (`.m3u8` file) which will enable muiltiple qualities and streaming. In addition, you can use their [API to get a VTT thumbnails file](https://docs.mux.com/guides/create-timeline-hover-previews#webvtt) that can be provided straight to the player.

>  If I'd like to use MUX, do I have to move all my videos there?

Yes, unless you want to also keep the original file copies in DO but that would cost more.

**Reply by piszczu4** [reply]
Mux looks very promising indeed :) Just one extra question. I started to google some solution and [here](https://www.reddit.com/r/aws/comments/wjol2j/is_cloudflare_r2_the_best_service_to_use_for/) one user mention the following solution:

> Just use S3 + CloudFront + MediaConvert. Your workflow sounds like a pretty common VOD workflow, that is exactly what MediaConvert is for.


I'm really newbie but just wanted to know which solution from your perspective would better match my needs, i.e. this one or MUX? I'm gonna store plenty of 10 minutes videos (educational ones).

**Reply by mihar-22** [reply]
I can't say definitively without knowing more about the types of videos, number of videos and durations, views, and site analytics on user area network/devices, but in general... streaming using Mux is going to deliver a much better experience for you and your users then using a standard AWS workflow. Setting these VOD workflows up and managing them can be painful, my advice is just leave it to the pros.

With respect to pricing, you'll end up paying encoding fees through MediaConvert, you'll pay storage fees for all of your video renditions (480p, 720p, 1080p, ...), and depending on the region and video size delivered varying bandwidth costs with CloudFront. In comparison, Mux encoding is free up to 1080p, uses just-in-time encoding to be super fast, storage is charged only per minute of the original video file, and streaming across a multi-CDN for the best possible performance is charged at a fixed rate regardless of size of delivery. In my opinon, just go with Mux and focus on providing awesome educational content :)

**Reply by piszczu4** [reply]
Convinced, Thank you!

**Reply by piszczu4** [reply]
I noticed that [here](https://github.com/vidstack/player/discussions/935) you wrote

> Expanding beyond the player, our entire focus at Vidstack is on: (1) ensuring top-notch performance across the CMS, player, and storage/encoding/streaming, (2) extremely fair pricing, (3) a seamless start-to-finish workflow by integrating with dev tools/frameworks. Lastly, I'm working on a partnership for our encoding/streaming, so if all goes well then I'm super excited to announce that hopefully very soon 🤞


You mean that you are gonna have your own service for video storing and streaming like MUX? Will it be comparable in terma of features/pricing?

**Reply by mihar-22** [reply]
Yes, we’ll be providing these services (packaged for the frontend) and partnering with a company to make it happen. I can’t say too much right now but we’ll announce stuff before end of year.
